src/main/java/org/commcare/core/process/CommCareInstanceInitializer.java

```diff
@@ -89,10 +90,35 @@ public InstanceRoot generateRoot(ExternalDataInstance instance) {
             return setupRemoteData(instance);
         } else if (ref.contains("migration")) {
             return setupMigrationData(instance);
+        } else if (ref.contains("selected_cases")) {
+            return setupSelectedCases(instance);
         }
         return ConcreteInstanceRoot.NULL;
     }
 
+    /**
+     * Initialises instances with reference to 'selected_cases'. Platforms that want to support these instances
+     * should extend this class to provide the implementation for this method
+     *
+     * @param instance Selected Cases Instance that needs to be initialised
+     * @return Initialised instance root for the the given instance
+     */
+    protected InstanceRoot setupSelectedCases(ExternalDataInstance instance) {
+        throw new RuntimeException(
+                "Instances with reference to 'selected_cases' are not supported on this platform");
+    }
+
+    @Nullable
+    protected String getGuidForSelectedCasesInstance(ExternalDataInstance instance) {
+        for (StackFrameStep step : session.getFrame().getSteps()) {
+            if (step.getId().equals(instance.getInstanceId()) && step.getType().equals(
+                    SessionFrame.STATE_MULTIPLE_DATUM_VAL)) {
+                return step.getValue();
```

Comment:
I was just looking at this code again and realised we're taking a different approach with storing this instance than we are with the results instance from query datums (where we put the instance source directly into the step).

I'm curious if we couldn't use the same mechanism for this?

Reply:
Hmm Good point, I didn't think about it until now and think we could use the same mechanism without any downsides. Although it seems like the wiring around that is pretty specific to `ExternalDataInstance` and `ExternalDataInstanceSource`. So we probably will need to make `VirtualDataInstance` an implementation of `ExternalDataInstance` and make a new type of `ExternalDataInstanceSource` which is more DB based rather than HTTP based.

Reply:
Yes that's all true but I think it's better design for the long run. I don't think it's necessary to do immediately but I expect we'd need to then do it for the 'case search input' work.

Reply:
yeah that makes sense. Since I am on PTO soon I will hold it off for now and we can revisit it next sprint and see whether to include it with case search input or multi-select work.

Reply:
@snopoke I think I have time this week to do this work though it seems like you have already started on 'case search input' work on FP which also touches this area. You have any suggestions on how to best manage the interlap between the two. I can start on it today and probably PR by tomorrow EOD but if you wanna take it with 'case search input' work that's also fine with me.

Reply:
@shubham1g5 I haven't gotten to that point yet. I'm currently just working on making it possible to have an entry with a post request work. So it should be fine for you to go ahead.

Reply:
Strongly agree with Simon that these two mechanisms seem very similar in structure and requirements and that it would be great to see the caching and loading unified. 

I think there's still some question in my mind about schema management expectations. Since we're defining the structure of these virtual docs, I think we'll want quite clear documentation about what the different kinds of virtual docs consist of, and we'll want to confirm whether we are going to provide versioning or graceful back/forwards compatibility with the types of changes we expect.

Reply:
>  we'll want to confirm whether we are going to provide versioning or graceful back/forwards compatibility

That's a good point, I would assume that we need to be backward compatible here in general. Possibly a good idea to add versions to the virtual docs right away to facilitiate smoother migrations in future ?

Corresponding changes are in https://github.com/dimagi/commcare-core/pull/1109